Firestore/Source/Model/FSTFieldValue.mm

```diff
@@ -1027,4 +959,114 @@ - (NSComparisonResult)compare:(FSTFieldValue *)other {
 
 @end
 
+@interface FSTDelegateValue ()
+@property(nonatomic, assign, readonly) FieldValue internalValue;
+@end
+
+@implementation FSTDelegateValue
++ (instancetype)delegateValue:(FieldValue &&)value {
+  return [[FSTDelegateValue alloc] initWithValue:std::move(value)];
+}
+
+- (id)initWithValue:(FieldValue &&)value {
+  self = [super init];
+  if (self) {
+    _internalValue = std::move(value);
+  }
+  return self;
+}
+
+- (FieldValue::Type)type {
+  return self.internalValue.type();
+}
+
+- (FSTTypeOrder)typeOrder {
+  switch (self.internalValue.type()) {
+    case FieldValue::Type::Null:
+      return FSTTypeOrderNull;
```

Comment:
Doesn't have to be in this PR, but `FSTTypeOrder` is almost replaceable with `FieldValue::Type`. The only case where it doesn't work exactly is in cases where typeOrder is compared for equality, and there we already have `FieldValue::Comparable`. The other case in `defaultCompare:` where we check for inequality should work as-is because comparable types are adjacent and the caller guarantees that it handles comparable types.